sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala

```diff
@@ -96,9 +103,37 @@ case class SortMergeJoin(
           RowIterator.fromScala(rightIter),
           numRightRows
         )
-        private[this] val joinRow = new JoinedRow
+
+        private[this] val rowJoiner = new JoinedRow
+        private[this] val unsafeRowJoiner =
+          GenerateUnsafeRowJoiner.create(leftSchema, rightSchema)
+
+        private[this] var joinedRow: InternalRow = _
+        private[this] val joinRow: (InternalRow, InternalRow) => InternalRow = {
+          if (canUnsafeJoin) {
+            (currentLeftRow: InternalRow, currentRightRow: InternalRow) => {
+              unsafeRowJoiner.join(currentLeftRow.asInstanceOf[UnsafeRow],
+                currentRightRow.asInstanceOf[UnsafeRow])
+            }
+          } else {
+            rowJoiner.apply
+          }
+        }
+
+        private[this] val withRight: (InternalRow) => InternalRow = {
+          if (canUnsafeJoin) {
+            (currentRightRow: InternalRow) => {
+              unsafeRowJoiner.withRight(currentRightRow.asInstanceOf[UnsafeRow])
+            }
+          } else {
+            rowJoiner.withRight
+          }
+        }
+
         private[this] val resultProjection: (InternalRow) => InternalRow = {
-          if (isUnsafeMode) {
+          if (canUnsafeJoin) {
+            identity[InternalRow]
```

Comment:
Two `UnsafeRows` are joined together as another `UnsafeRow`. So we don't need to do `UnsafeProjection` here.